### PR TITLE
Revised handling of the list of all opened channels

### PR DESCRIPTION
**Pull request by xavierleroy**
This is a variant on @Octachron's PR about I/O channels.

It simplifies the code by not putting channels opened from C on the list of all opened channels, only those opened from OCaml
  and tracked by the GC.

It also fixes an earlier error in `caml_finalize_channel` where the channel could be unlinked from the list, then not freed because not flushed.  But it will never be flushed during `at_exit` since it was unlinked !

### CODE REVIEW

File: runtime/io.c [hunk 1]
```diff
@@ -636,7 +625,7 @@ CAMLprim value caml_ml_out_channels_list (value unit)
     if (channel->max == NULL
         && channel->flags & CHANNEL_FLAG_MANAGED_BY_GC) {
```
Comment (Octachron) [inline runtime/io.c]: Shouldn't we remove this part of the test or replace it by an assert now that we only link gc-managed channels?

**Comment by Octachron** [comment]
Thanks! I was visibly missing a forest behind the tree this time. I will review the changes properly and probably merge them in this PR branch.

File: runtime/io.c (hunk 1 above)
Comment (xavierleroy) [reply runtime/io.c]: Yes, absolutely.  Will put an assertion.

**Comment by xavierleroy** [comment]
> I was visibly missing a forest behind the tree this time.

Well, this code is complicated to begin with, and went through (too?) many iterations and incremental changes already.

I considered putting output channels only in the doubly-linked list used by `caml_ml_out_channels_list`.  Input channels don't really need to be recorded there...  But this can be done later if we agree it's a good idea.

**Pull request merged**